packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts

```diff
@@ -232,3 +243,24 @@ export const getSpaceKeyFromDoc = (doc: any): string | null => {
 
   return String(rawSpaceKey);
 };
+
+const waitForHeads = async (handle: DocHandle<SpaceDoc>, heads: string[]) => {
+  await handle.whenReady();
+  if (headsArePresentInDoc(handle.docSync(), heads)) {
+    return;
+  }
+  // Wait for the document to receive one change it will kick it in sync with the thin-client.
+  const waitForChange = new Trigger();
+  const listener = ({ doc }: DocHandleChangePayload<SpaceDoc>) => {
+    if (headsArePresentInDoc(handle.docSync(), heads)) {
+      waitForChange.wake();
+      handle.off('change', listener);
+    }
+  };
+  handle.on('change', listener);
+  await waitForChange.wait({ timeout: FLUSH_TIMEOUT });
+};
+
+const headsArePresentInDoc = (doc: Doc<any>, heads: string[]): boolean => {
+  return heads.map((head) => getBackend(doc).getChangeByHash(head)).every((change) => change !== null);
```

Comment:
can you please cache those changes by hash in `waitForHeads` there might be significant overhead copying data from wasm to js